File: sdk-api/src/test/java/dev/restate/sdk/EagerStateTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import static dev.restate.sdk.JavaBlockingTests.testDefinitionForVirtualObject;
import static org.assertj.core.api.Assertions.assertThat;

import dev.restate.sdk.core.EagerStateTestSuite;
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder;
import dev.restate.sdk.core.TestSerdes;
import dev.restate.sdk.types.StateKey;
import dev.restate.serde.Serde;

public class EagerStateTest extends EagerStateTestSuite {

  @Override
  protected TestInvocationBuilder getEmpty() {
    return testDefinitionForVirtualObject(
        "GetEmpty",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, unused) ->
            String.valueOf(ctx.get(StateKey.of("STATE", TestSerdes.STRING)).isEmpty()));
  }

  @Override
  protected TestInvocationBuilder get() {
    return testDefinitionForVirtualObject(
        "GetEmpty",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, unused) -> ctx.get(StateKey.of("STATE", TestSerdes.STRING)).get());
  }

  @Override
  protected TestInvocationBuilder getAppendAndGet() {
    return testDefinitionForVirtualObject(
        "GetAppendAndGet",
        TestSerdes.STRING,
        TestSerdes.STRING,
        (ctx, input) -> {
          String oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING)).get();
          ctx.set(StateKey.of("STATE", TestSerdes.STRING), oldState + input);

          return ctx.get(StateKey.of("STATE", TestSerdes.STRING)).get();
        });
  }

  @Override
  protected TestInvocationBuilder getClearAndGet() {
    return testDefinitionForVirtualObject(
        "GetClearAndGet",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, input) -> {
          String oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING)).get();

          ctx.clear(StateKey.of("STATE", TestSerdes.STRING));
          assertThat(ctx.get(StateKey.of("STATE", TestSerdes.STRING))).isEmpty();
          return oldState;
        });
  }

  @Override
  protected TestInvocationBuilder getClearAllAndGet() {
    return testDefinitionForVirtualObject(
        "GetClearAllAndGet",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, input) -> {
          String oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING)).get();

          ctx.clearAll();
          assertThat(ctx.get(StateKey.of("STATE", TestSerdes.STRING))).isEmpty();
          assertThat(ctx.get(StateKey.of("ANOTHER_STATE", TestSerdes.STRING))).isEmpty();

          return oldState;
        });
  }

  @Override
  protected TestInvocationBuilder listKeys() {
    return testDefinitionForVirtualObject(
        "ListKeys",
        Serde.VOID,
        TestSerdes.STRING,
        (ctx, input) -> String.join(",", ctx.stateKeys()));
  }

  @Override
  protected TestInvocationBuilder consecutiveGetWithEmpty() {
    return testDefinitionForVirtualObject(
        "ConsecutiveGetWithEmpty",
        Serde.VOID,
        Serde.VOID,
        (ctx, input) -> {
          assertThat(ctx.get(StateKey.of("key-0", TestSerdes.STRING))).isEmpty();
          assertThat(ctx.get(StateKey.of("key-0", TestSerdes.STRING))).isEmpty();
          return null;
        });
  }
}

File: sdk-core/bin/test/dev/restate/sdk/core/kotlinapi/EagerStateTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.core.kotlinapi

import dev.restate.sdk.common.StateKey
import dev.restate.sdk.core.EagerStateTestSuite
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder
import dev.restate.sdk.core.TestSerdes
import dev.restate.sdk.core.kotlinapi.KotlinAPITests.Companion.testDefinitionForVirtualObject
import org.assertj.core.api.AssertionsForClassTypes.assertThat

class EagerStateTest : EagerStateTestSuite() {
  override fun getEmpty(): TestInvocationBuilder =
      testDefinitionForVirtualObject("GetEmpty") { ctx, _: Unit ->
        val stateIsEmpty = ctx.get(StateKey.of("STATE", TestSerdes.STRING)) == null
        stateIsEmpty.toString()
      }

  override fun get(): TestInvocationBuilder =
      testDefinitionForVirtualObject("GetEmpty") { ctx, _: Unit ->
        ctx.get(StateKey.of("STATE", TestSerdes.STRING))!!
      }

  override fun getAppendAndGet(): TestInvocationBuilder =
      testDefinitionForVirtualObject("GetAppendAndGet") { ctx, name: String ->
        val oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING))!!
        ctx.set(StateKey.of("STATE", TestSerdes.STRING), oldState + name)
        ctx.get(StateKey.of("STATE", TestSerdes.STRING))!!
      }

  override fun getClearAndGet(): TestInvocationBuilder =
      testDefinitionForVirtualObject("GetClearAndGet") { ctx, _: Unit ->
        val oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING))!!
        ctx.clear(StateKey.of("STATE", TestSerdes.STRING))
        assertThat(ctx.get(StateKey.of("STATE", TestSerdes.STRING))).isNull()
        oldState
      }

  override fun getClearAllAndGet(): TestInvocationBuilder =
      testDefinitionForVirtualObject("GetClearAllAndGet") { ctx, _: Unit ->
        val oldState = ctx.get(StateKey.of("STATE", TestSerdes.STRING))!!

        ctx.clearAll()

        assertThat(ctx.get(StateKey.of("STATE", TestSerdes.STRING))).isNull()
        assertThat(ctx.get(StateKey.of("ANOTHER_STATE", TestSerdes.STRING))).isNull()
        oldState
      }

  override fun listKeys(): TestInvocationBuilder =
      testDefinitionForVirtualObject("ListKeys") { ctx, _: Unit ->
        ctx.stateKeys().joinToString(separator = ",")
      }

  override fun consecutiveGetWithEmpty(): TestInvocationBuilder =
      testDefinitionForVirtualObject("ConsecutiveGetWithEmpty") { ctx, _: Unit ->
        assertThat(ctx.get(StateKey.of("key-0", TestSerdes.STRING))).isNull()
        assertThat(ctx.get(StateKey.of("key-0", TestSerdes.STRING))).isNull()
      }
}
